### PR TITLE
Add Brant: foundation model for intracranial (sEEG/iEEG) signals

### PR DESCRIPTION
**Pull request by adammounir**
## Summary

Towards #1097 — adds a Braindecode-native implementation of **Brant**
(Zhang et al., *Brant: Foundation Model for Intracranial Neural Signal*,
NeurIPS 2023) for intracranial EEG signals.

This PR is an **architecture-only contribution**. It does not ship, advertise,
or endorse a verified Brant pretrained checkpoint. Braindecode's generic
save/load API remains available for checkpoints created from this
implementation; verification and redistribution of third-party weights require
separate provenance, parity, and licensing gates.

## Included

- `Brant(EEGModuleMixin, nn.Module)` with in-model band-power features, patch
  tokenization, temporal and spatial Transformer encoders, and a prediction
  head.
- Standard Braindecode contracts for runtime signal geometry,
  `return_features`, `reset_head`, model registration, summary metadata, and
  serialized configuration.
- Reduced-precision-safe spectral computation with float32/float64 numerical
  coverage against SciPy.
- Unit, integration, export/compile, configuration, and local save/load tests.
- Model documentation, examples, a 1.8 changelog entry, and a conditional
  developer parity helper that requires explicit upstream weight mapping.
- The exact upstream Apache-2.0 license plus attribution and modification
  notices in source, sdist, and wheel distributions. Braindecode remains
  BSD-3-Clause overall.

## Architecture

`(B, C, T) -> patch + band-power embedding -> temporal Transformer -> spatial
Transformer -> prediction head`

Frequency-domain features are computed inside `forward`, preserving the
standard `(B, C, T)` model input. The spatial path is channel-agnostic; temporal
geometry is validated against the configured sample length.

## Validation

- Focused Brant and feature-contract tests: 24 passed, 1 gated-source skip.
- Integration tests: 11 passed, 1 documented TorchScript skip.
- Model registration/summary sweep: 170 passed, 4 skips.
- Configuration/Hub-contract tests, documentation-link checks, package builds,
  and all changed-file pre-commit hooks pass locally.
- Fresh CI is required on the exact updated PR head.

## Upstream and scope boundary

- Source: https://github.com/yzz673/Brant (Apache-2.0; audited revision recorded
  in `NOTICE.txt`).
- Paper: https://proceedings.neurips.cc/paper_files/paper/2023/file/535915d26859036410b0533804cee788-Paper-Conference.pdf
- No official or converted weight artifact is part of this PR's deliverables or
  parity claims.

## Checklist

- [x] Implement and register the architecture.
- [x] Enforce Braindecode model and feature contracts.
- [x] Add correctness, integration, configuration, and serialization coverage.
- [x] Package upstream license and modification attribution.
- [x] Document the architecture-only checkpoint boundary.
- [x] Obtain fresh green CI and close superseded review threads on this exact
      head.

### CODE REVIEW

**Comment by bruAristimunha** [comment]
FYI, @BelCorentin, @hubertjb, @teonbrooks, and @kbegany, @adammounir has started to implement the iEEG models list here, in case you need. 


https://github.com/braindecode/braindecode/issues/1097

**Comment by adammounir** [comment]
@bruAristimunha Brant looks good to me, happy to receive your review on it :)

**Comment by bruAristimunha** [comment]
hey @adammounir,

Where is the replicability?

**Comment by adammounir** [comment]
I fed identical inputs through the upstream encoders (`Brant_src/pretrain/pre_model.py`, Apache-2.0) and the braindecode port with the weights copied over. Both are **bit-exact**:

| Check | Upstream ref | braindecode port | max\|diff\| |
|---|---|---|---|
| Temporal encoder | `TimeEncoder` | `_BrantTemporalEncoder` | **0.0** |
| Spatial encoder | `ChannelEncoder` | `_BrantSpatialEncoder` | **0.0** |
| `from_pretrained` | `Daoze/Brant` (508M) | `braindecode/brant-pretrained` (508M) | loads + forward OK |

Reproduce with `scripts/brant_parity_check.py --brant-src <Brant_src>`; gated in CI by `test_brant.py::test_encoders_are_bit_exact_with_upstream` (skips without `BRANT_SRC`). Weights re-hosted at `braindecode/brant-pretrained` (Apache-2.0): `Brant.from_pretrained("braindecode/brant-pretrained", n_outputs=2)`.

File: braindecode/models/__init__.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 from .bendr import BENDR, InterpolatedBENDR
 from .biot import BIOT, InterpolatedBIOT
 from .brainmodule import BrainModule
+from .brant import Brant
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/models/__init__.py:10]: Reconciled in the live PR description: Brant is intentionally public, registered, present in `summary.csv`, and covered by the shared model contracts. Exact head `ea61da9de765ce037d1b438ebf0d5483cb980a73` passed all 13 checks. The older non-registration wording quoted here is no longer present.

File: braindecode/models/util.py [hunk 2]
```diff
@@ -445,6 +445,7 @@ def _get_bendr_chs_info() -> list[dict]:
     ("PBT", ["n_chans", "n_outputs", "n_times"], None),
     ("SSTDPN", ["n_chans", "n_outputs", "n_times", "sfreq"], None),
     ("BrainModule", ["n_chans", "n_outputs", "n_times", "sfreq"], None),
+    ("Brant", ["n_chans", "n_outputs", "n_times", "sfreq"], None),
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/models/util.py:448]: The current PR scope intentionally registers Brant and exercises it through the shared model suites. The live body at exact head `ea61da9de765ce037d1b438ebf0d5483cb980a73` now says this explicitly, and that head passed all 13 checks, including the six-platform matrix, docs, and pre-commit. This thread quotes the superseded draft description; no registration rollback is needed.

File: braindecode/models/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/models/__init__.py:10]: The current PR scope intentionally imports and registers Brant, includes its summary/mandatory-parameter metadata, and covers it in the shared suites. The live body at exact head `ea61da9de765ce037d1b438ebf0d5483cb980a73` now reflects that scope, and all 13 exact-head checks pass. This concern was against the superseded draft description.

File: braindecode/models/brant.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+        self.band_encoding = nn.Parameter(torch.randn(n_bands, d_model))
+        self.positional_encoding = nn.Parameter(torch.randn(seq_len, d_model))
+        self.proj = nn.Sequential(nn.Linear(patch_size, d_model))
+        self.softmax = nn.Softmax(dim=-1)
+
+    def forward(self, data: torch.Tensor, power: torch.Tensor) -> torch.Tensor:
+        """Embed raw patches together with frequency and position information."""
+        batch_size, n_chans, seq_len, patch_size = data.shape
+        power = self.softmax(power)
+        power_emb = torch.einsum("hijk, kl->hijl", power, self.band_encoding)
```
[the remark this answers is not in the document]
Comment (bruAristimunha) [reply braindecode/models/brant.py:371]: No code change is needed here. On the exact-head environment (PyTorch 2.13.0), `torch.einsum("hijk, kl->hijl", power, encoding)` executes successfully and is bit-identical to the no-space equation on the same tensors (shape `(2, 3, 4, 7)`). The focused Brant/integration gates and all 13 exact-head CI checks are also green; PyTorch accepts this equation whitespace.

**Review by bruAristimunha — CHANGES_REQUESTED** [review]
I reviewed this exact candidate against the primary Brant paper and upstream
artifacts. I am requesting changes because the required paper-result gate is
not yet executable, so this new model is not merge-qualified even if the
architecture tests are green.

The bounded qualifying target is Appendix G, Table 10, MAYO short-term
forecasting: MAE `0.6860`, MSE `0.8064`. The current evidence cannot identify
or reproduce that row:

- the versioned NEMAR BIDS mirror exposes 24 participants, the data paper
  reports 25 acquired patients, and the Brant paper reports 18 used subjects;
  no immutable manifest identifies those 18, the sampled 400 minutes, the
  random 320/80-minute split, or whether the paper used the continuous
  BIDS/MEF distribution rather than the separately published three-second
  segmented MATLAB clips;
- the official fine-tuning/evaluation source is access-gated, and the mutable
  checkpoint pointer has no accepted bytes, SHA-256, inventory, row lineage,
  or reconciled reuse terms;
- this PR exposes a pooled classification head, not the paper's per-channel
  future-signal forecasting head.

Please keep this held until an immutable protocol closes those identities and
one exact candidate SHA reproduces the selected row within a preregistered
tolerance. The smallest reviewable shape is a compact representation-only
architecture with an explicit official-checkpoint/paper hold in the library
and an external opt-in reproduction lane; official checkpoint
loading/translation and broad benchmark/data machinery should not be added
until immutable source and checkpoint provenance is qualified. No MAYO or
checkpoint bytes were downloaded because the scientific and provenance gates
are incomplete.

Reviewed head: `ea61da9de765ce037d1b438ebf0d5483cb980a73`.

**Comment by bruAristimunha** [comment]
Paper-replication gate at exact head ea61da9de765ce037d1b438ebf0d5483cb980a73: hold.

NEMAR nm000182 (MAYO) and nm000183 (FNUSA) are identity-preserving public conversions, but their objects are independent 3-second single-channel clips. They cannot supply Brant's simultaneous-channel 6-second patches, 90-second history, or 12/120-second future targets; concatenating them would fabricate the paper input. The paper also does not publish the selected Appendix-G subject IDs or seizure-label mapping.

A fail-closed admissibility audit passed and documents this impossibility. Merge still requires accepted immutable upstream source/weights, source-to-port numerical parity, redistribution terms, and an admissible paper-task result. CI/architecture smoke alone is not sufficient.
